apps/labs/posts/numpy-string-ufuncs.md

```diff
@@ -0,0 +1,546 @@
+                                    npy_intp *view_offset)
+{
+    loop_descrs[0] = NPY_DT_CALL_ensure_canonical(given_descrs[0]);
+    if (loop_descrs[0] == NULL) {
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+
+    loop_descrs[1] = NPY_DT_CALL_ensure_canonical(given_descrs[1]);
+    if (loop_descrs[1] == NULL) {
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+
+    // Create a new descriptor instance from the first input and add
+    // the element size of the second input.
+    loop_descrs[2] = PyArray_DescrNew(loop_descrs[0]);
+    if (loop_descrs[2] == NULL) {
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+    loop_descrs[2]->elsize += loop_descrs[1]->elsize;
+
+    return NPY_NO_CASTING;
+}
+```
+
+It does the following:
+- Specifies that the input dtypes are the ones that should be
+  used for passing information to the loop. In doing so, it also ensures that
+  the dtypes are in canonical representation, meaning in the native byte order of
+  the underlying system.
+- For the output dtype (`loop_descrs[2]`), we create a new descriptor instance
+  from one of the inputs (either one would be fine), and then add the element size
+  of the other one. This way, we get an output dtype with a size that's equal to
+  the sum of the sizes of the two input dtypes.
+
+After registering the `resolve_descriptors` function, everything works correctly
+and `np.add` now works for string dtypes as well!
+
+## `find` / `rfind`: Nothing new here, right?
+
+Next in our list of functions to implement as ufuncs was `find` and `rfind`. Like
+`isalpha`, these two return fixed-type dtypes (`int`s), so no need for a
+`resolve_descriptors` function, and they both would be new ufuncs. Unlike
+`isalpha` though, they both accept four input arguments: the buffer to search in,
+the substring to search for, and starting and ending indices (if one wants to
+specify in what part of the buffer to search). The starting and ending indices
+can be of any integer dtype (int8, int16, int32 etc.), but we only want to write
+one loop for all of them. To achieve that we need a promoter.
+
+A promoter is a function that specifies the promotion rules for a ufunc. In general,
+a ufunc will search for a loop that matches the input dtypes exactly. In order
+to be able to accept more than one input dtype, and have all of those dispatch
+to the same loop, we should register a promoter. A promoter has the following
+signature:
+
+```cpp
+int promoter(PyUFuncObject *ufunc,
+             PyArray_DTypeMeta **op_dtypes,
+             PyArray_DTypeMeta **signature,
+             PyArray_DTypeMeta **new_op_dtypes)
+```
+
+The two important arguments here are `op_dtypes` and `new_op_dtypes`. Similar
+to the `resolve_descriptors` function, `op_dtypes` are the dtypes the user passed
+into the ufunc. `new_op_dtypes` is an array of dtypes that must be filled with
+the dtypes the input should be promoted to before searching for a loop.
```

Comment:
This is better documentation than what is currently in the numpy docs! Consider improving those docs if you have the time and motivation. https://numpy.org/devdocs/reference/c-api/array.html#c.PyArrayMethod_PromoterFunction

Reply:
Will do!